### PR TITLE
Integrate flake8_rst into ./ci/code_check.sh

### LINKED ISSUE #23154: Validate PEP-8 in docstring examples
At the moment we are automatically validating PEP-8 in our code (using `flake8 .` in the CI), and we also have the script `scripts/validate_docstrings.py` that reports errors for different formatting errors in our docstrings, and it also runs the tests to see if they work and generate the expected input. But we are not validating whether the examples follow PEP-8.

In the next example:
```
def head():
    """
    Return the first n rows of the Series.

    Examples
    --------
    >>> s=pd.Series([1, 2, 3, 4])
    >>> s.head(2)
    0    1
    1    2
    dtype: int64
```

Everything will be reported as correct, because the formatting of the docstring is the expected, and when running the examples, they work and they generated the presented output.

According to PEP-8, there should be spaces around the `=` in an assignment. So, we should have `s = pd.Series([1, 2, 3, 4])` instead of `s=pd.Series([1, 2, 3, 4])`. So, the validation should report it as an error.

What we need to do is:
- Add a method in `scripts/validate_docstrings.py` that obtains the code in the examples (and if possible their line in the code to be presented to the user if there are errors). For what I know, the module `doctest` in the Python standard library provides a way to extract the code.
- Add another method that given some code, returns possible PEP-8 errors. We currently have `pyflakes` as a dependency, so using it is probably the best option. But if it's better to use `pycodestyle` or something else, that could be an option.
- Use these methods in the validation, so errors are reported as the rest of the errors from the validation
- Add unit tests for the implemented features in `scripts/tests/test_validate_docstrings.py`

### PR DESCRIPTION
**Pull request by FHaase**
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

Integrated [flake8-rst](https://github.com/kataev/flake8-rst) into `code_check.sh` in order to check code in .. code-blocks:: python directives

* fixed all issues found with version v0.4.1
    - F821 undefined name ignored  -- too many occurences for now
    - W391 blank line at end of file  -- 4 occurences where I had no idea how to fix

```
doc/source/contributing_docstring.rst:629:9: W391 blank line at end of file
doc/source/contributing_docstring.rst:725:9: W391 blank line at end of file
doc/source/contributing_docstring.rst:815:9: W391 blank line at end of file
doc/source/contributing_docstring.rst:911:9: W391 blank line at end of file
```

### CODE REVIEW

**Comment by gfyoung** [comment]
> F821 undefined name ignored -- too many occurences for now

If this gets merged, an issue should be created for this.

**Comment by datapythonista** [comment]
Really nice work, thank you for it.

But I don't think this is really validating the docstrings, just the rst files. We surely have plenty of pep8 errors in docstrings that are not detected here.

Besides that, in #23154, the goal was to validate pep8 of docstrings from `validate_docstrings.py`, so people working on docstrings can see of there is any pep8 issue at the same time of validating the rest of formats.

Happy to merge this, but let's remove the "Closes #23154" from the description, and let's create the issues for the pep8 errors being ignored as @gfyoung says.

Also, can't check well from my phone, but when adding flake8-rst to the dependencies, there are some files that are automatically generated. Did you edit them manually, or did you run `convert_deps.py`?


@jorisvandenbossche, @WillAyd, if you can also take a look here.

**Comment by FHaase** [comment]
Okay let me summarize what I did:
* added flake8-rst to `./ci/environment-dev.yaml` then executed `./scripts/convert_deps.py
   - it changed requirement `tables` within `./ci/requirements-optional-pip.txt` to something with `pytables`.
        manually reverted back to `tables` as `pytables` does not seem to be within [pypi.org](https://pypi.org/search/?q=pytables)
* added flake8-rst commands to `./ci/code_checks.sh` 
   - flake8-rst pandas                       # searching within *.py files in pandas subfolder. 
   - flake8-rst doc --filename=*.rst   # searching within *.rst files in doc subfolder
* fixed all issues flake8-rst in version v0.4.1 finds executing both commands.
* added flake8-rst as dependency in `./ci/travis-36.yaml` to make travis pass.

What does flake8-rst do?
* looks for `.. code-blocks:: python` directives 
* cuts the code-block out and passes it into flake8
* reports the flake8 issues

Does it find everything? 
* No
   - `::` or `.. code-blocks::`                 # No specification of language python
   - `.. ipython:: python`                      # ipython currently not in RegEx 
   - `flake8-rst pandas --benchmark`  # reveals number of checked files is 3
* The intention of the plugin however is to find all these things.

Regarding integration in `validate_docstrings.py`: 
I didn't get it was specificly required to have it integrated there and thought having it in `./ci/code_checks.sh` would already be enough. **I'm going to take a look into that.**

Why do I think it closes #23154 ?
The PR introduces a mechanic that [if bugless] checks code in docstrings and rst files. Currently it isn't bugfree but that is not an issue with this project, but rather with flake8-rst itself and should be fixed over there.

**Comment by FHaase** [comment]
@datapythonista 
calling `./ci/validate_docstrings.py` now calls and outputs results of:
* `flake8 --doctest <doc.source_file_name>?`
* `flake8-rst <doc.source_file_name>?`
`<doc.source_file_name>` only if a function name was passed.

`./ci/validate_docstrings.py pandas.read_excel` 
```
Flake8 reported issues:
Invoking command "flake8 --doctest pandas/util/_decorators.py" returned:
pandas/util/_decorators.py:320:26: F721 syntax error in doctest
pandas/util/_decorators.py:321:13: F721 syntax error in doctest
pandas/util/_decorators.py:322:15: F821 undefined name '_make_signature'
pandas/util/_decorators.py:322:31: F821 undefined name 'f'
```
Is that what you had in mind?

**Comment by datapythonista** [comment]
As I said, I was happy to merge the PR as it was. Just, we can't close the issue you referenced yet, as there is more to do.

If you can add those new changes in a separate PR I can give you more precise feedback. But summarizing, `validate_doctrings.py` is reporting a set of errors of different types, and PEP8 issues are just another kind of error. I don't like how flake8 is called as a new type of checked, and not as the rest of the checks. This makes things much more complicated, and doesn't consider the pep8 errors when the script is called without parameters to get the json report of all pending errors in the docstrings.

Besides that, I don't like using subprocess. I guess flake8-rst is a Python program, so we should be able to import it and call the appropriate functions.

In any case, validating the docstrings is for another PR, as it's independent of checking the code blocks.

File: doc/source/basics.rst
Comment (WillAyd) [inline doc/source/basics.rst]: What was the point of changing this? We always ask changes to have one focus to keep diff minimal. Would prefer if you could back out anything unrelated to the flake8 and do in a separate PR

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: doc/source/contributing_docstring.rst
Comment (FHaase) [reply doc/source/contributing_docstring.rst]: its integrated directly in [sphinx](https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html?highlight=pep#role-pep)

File: doc/source/basics.rst
Comment (FHaase) [reply doc/source/basics.rst]: :+1: didn't know that directive.

File: doc/source/basics.rst
Comment (FHaase) [reply doc/source/basics.rst]: I guess the `and` tricked me into thinking it was an if-statement like a couple of lines above.

File: doc/source/contributing_docstring.rst
Comment (FHaase) [reply doc/source/contributing_docstring.rst]: At that position no, I went for symmetry. Indeed using `noqa` is quite confusing and therefore I try to avoid it. 
@WillAyd At this position I thought putting every method in an extra .. code-block:: would be the best option, however it changes the documentation quite signifikant?!

File: doc/source/contributing.rst
Comment (FHaase) [inline doc/source/contributing.rst]: Would you prefer a `pass` instead of the `noqa`?

**Review by datapythonista — APPROVED** [review]
Really nice work. I'd possibly use `pass` instead of the `...` everywhere, but just a preference.

Looking forward to have all these fixes merged, and to start validating pep8 in docstrings.

Btw, if you have rendered the output of the `pep` directive and you can add a screenshot of it, would be nice, I don't know how that is being rendered.

**Comment by FHaase** [comment]
pep257
pep8

Found another one when looking for the occurence.

**Comment by datapythonista** [comment]
doesn't seem like the rendering is correct, is it? I guess what's expected is that the text shown is `PEP 8` and that is a link to the pep8 page. Can't check the links, but looks like there are two different links there?

Sorry if I'm misunderstanding.

**Comment by FHaase** [comment]
`<a class="pep reference external" href="https://www.python.org/dev/peps/pep-0008#imports"><strong>PEP 8#imports</strong></a>`
That's the generated html. It looks a little strange because of the linebreak. The link works and points to the import section of the pep page.

I'd rather use the `:pep:` role than having a hardcoded link everywhere.

> **:pep:**
>    A reference to a Python Enhancement Proposal. This generates appropriate index entries. The text “PEP number” is generated; in the HTML output, this text is a hyperlink to an online copy of the specified PEP. You can link to a specific section by saying :pep:\`number#anchor\`.

**Comment by datapythonista** [comment]
thanks for the clarification, if that's the case all good.

@WillAyd can you take a look and merge if you're happy?

**Comment by FHaase** [comment]
Before merging I'd like to change two more things:
* remove flake8_rst coverage from .py files as there won't be any `.. code-blocks::` within NumpyDocstrings. And flake8 with --docstrings options will cover it once #23399 is merged and integrated in ci.
* temporarily fix version to `0.4.2` as newer version will cover more types ( especially `.. ipython::`), thus suddenly letting travis fail.

I'm just waiting for any more suggestions to not trigger ci-builds. [skip ci] seems to not work with circle ci and azure pipeline.

**Comment by datapythonista** [comment]
I wouldn't worry that much about the CI, and I'd leave the PR in a state that you're happy and it can be merged.

**Comment by FHaase** [comment]
Okay can merge if wanted

**Comment by datapythonista** [comment]
@WillAyd can you take another look, and merge on green if you're happy.

Thanks @FHaase for the work on this, really nice PR. If you want to create two issues for the errors you're ignoring, that would be nice. I'm sure you'll be better than me at explaining what are the problems. You can mention this PR, and mention me too if you want.

**Comment by jreback** [comment]
@datapythonista note this will slightly conflict with your revision of ci deps (need to add the new package)

**Comment by datapythonista** [comment]
no problem, I can update the other two PRs, their CI is still waiting, so it's a good time

**Comment by datapythonista** [comment]
@FHaase can you fix the conflicts please? Sorry this wasn't merged before.

**Comment by FHaase** [comment]
@datapythonista No Problem. Currently fixing issues within the docs getting covered by an updated flake8-rst. Is it possible to add a PR that builds on an unmerged PR?

**Comment by datapythonista** [comment]
It's possible, but makes things a bit tricky, so I'd better wait.

@jreback @WillAyd can you take a look at this again? Would be nice to get it merged once it's green.

**Review by jorisvandenbossche — APPROVED** [review]
Looks good to me!

**Pull request merged**
